### PR TITLE
rework msgpack buffers

### PR DESCRIPTION
**Pull request by richardstartin**
The main change here is that buffering is abstracted away from the format writer(s?) so that all the format writer needs to be aware of is the format. There is a new abstraction called `StreamingBuffer` with two flavours
* `FlushingBuffer` - fixed size, flushes when full, this is how we have been sending traces for months. Traces are awkward because they are large and the format is length prefixed, which means we need to tell the agent how many traces to expect before sending the traces, which forbids streaming.
* `GrowableBuffer` - resizes its buffer when necessary, should only be used when there is an implicit limit on its growth in practice, e.g. for the serialized string table in the v0.5 trace format.

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java [hunk 1]
```diff
@@ -37,10 +34,7 @@
 
   public SerializingMetricWriter(WellKnownTags wellKnownTags, Sink sink) {
     this.wellKnownTags = wellKnownTags;
-    this.writer =
-        new MsgPackWriter(
-            // 64KB
-            sink, ByteBuffer.allocate(64 << 10), EnumSet.of(RESIZEABLE, SINGLE_MESSAGE));
+    this.writer = new MsgPackWriter(new FlushingBuffer(5 << 20, sink));
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java:37]: Will be replaced with a buffer over an okhttp `BufferedSink` obtained from `Pipe.source()` now that the flushing logic is moved out of the `MsgPackWriter` - this will allow metrics to be piped through a tiny 8KB buffer with an asynchronous sender, reducing footprint and avoid resizes or over-sizing.

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java [hunk 2]
```diff
@@ -25,4 +29,24 @@ int traceCount() {
   abstract void writeTo(WritableByteChannel channel) throws IOException;
 
   abstract RequestBody toRequest();
+
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java:32]: the bleeding of msgpack into here seems a reasonable tradeoff for not needing to store the header in the buffer itself, which makes `MsgPackWriter` more generic.

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/msgpack/MsgPackWriter.java [hunk 3]
```diff
@@ -138,80 +147,12 @@ public void writeString(CharSequence s, EncodingCache encodingCache) {
...
-      // doing it yourself (avoids lots of bounds checks).
-      // Using UTF8ByteString wherever possible should make this rare anyway.
-      byte[] utf8 = ((String) s).getBytes(UTF_8);
-      buffer.put(utf8);
-      return utf8.length;
-    }
-    return allocationFreeUTF8Encode(s);
-  }
-
-  private int allocationFreeUTF8Encode(CharSequence s) {
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/msgpack/MsgPackWriter.java:175]: `UTF8BytesString` obviated the need for these sorts of shenanigans, which are actually a lot slower than calling `getBytes()` because every `charAt()` and every `put(byte)` is bounds-checked. The more we use `UTF8BytesString`, the less we'll allocate in serialisation.

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/WritableFormatter.java [hunk 4]
```diff
@@ -1,150 +1,3 @@
 package datadog.trace.core.serialization;
 
-import datadog.trace.bootstrap.instrumentation.api.UTF8BytesString;
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/WritableFormatter.java:3]: This existed to share flushing logic between the abandoned protobuf writer and the msgpack writer. We won't be trying protobuf again, and the flushing has been moved into the buffer implementation.

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentApiTest.groovy [hunk 5]
```diff
@@ -391,9 +392,8 @@ class DDAgentApiTest extends DDSpecification {
   }
 
   Payload prepareTraces(String agentVersion, List<List<DDSpan>> traces) {
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentApiTest.groovy:394]: This sort of thing is in too many places and made this change too difficult. I've patched up all of these, but this needs refactoring.

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy [hunk 6]
```diff
@@ -200,7 +201,7 @@ class DDAgentWriterCombinedTest extends DDSpecification {
     when:
     def mapper = agentVersion.equals("v0.5/traces") ? new TraceMapperV0_5() : new TraceMapperV0_4()
     int traceSize = calculateSize(minimalTrace, mapper)
-    int maxedPayloadTraceCount = ((int) ((mapper.messageBufferSize() - 5) / traceSize))
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy:203]: 5 was the space reserved for a header.

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV04PayloadTest.groovy [hunk 7]
```diff
@@ -204,6 +205,13 @@ class TraceMapperV04PayloadTest extends DDSpecification {
 
     @Override
     int write(ByteBuffer src) {
+      if (captured.remaining() < src.remaining()) {
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV04PayloadTest.groovy:208]: letting this grow allows the test to explore more cases, rather than rejecting the output, but the increase in heap usage needs to be monitored.

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV05PayloadTest.groovy [hunk 8]
```diff
@@ -32,49 +33,6 @@ import static org.msgpack.core.MessageFormat.UINT8
 
 class TraceMapperV05PayloadTest extends DDSpecification {
 
-
-  def "dictionary overflow causes a flush"() {
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV05PayloadTest.groovy:36]: The dictionary doesn't flush any more. Keeping two isolated callbacks triggered when fixed capacities were reached in sync made this code extremely hard to reason about.

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV05PayloadTest.groovy [hunk 9]
```diff
@@ -98,7 +56,9 @@ class TraceMapperV05PayloadTest extends DDSpecification {
       UUID.randomUUID().toString(),
       false))
     int traceSize = calculateSize(repeatedTrace)
-    int tracesRequiredToOverflowBody = (traceMapper.messageBufferSize() + traceSize - 1) / traceSize
+    // 30KB body
```
Comment (richardstartin) [inline dd-trace-core/src/test/groovy/datadog/trace/common/writer/ddagent/TraceMapperV05PayloadTest.groovy:59]: no need to use MBs of data to check there's a flush when the message buffer is full

File: internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/UTF8BytesString.java [hunk 10]
```diff
@@ -53,6 +53,12 @@ public void transferTo(ByteBuffer buffer) {
     buffer.put(utf8Bytes);
   }
 
+  /** Writes the UTF8 encoding of the wrapped {@code String}. */
+  public byte[] getUtf8Bytes() {
```
Comment (richardstartin) [inline internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/UTF8BytesString.java:57]: Ideally a naked reference to this `byte[]` should be avoided

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java (hunk 2 above)
Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java:32]: 👍

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/msgpack/MsgPackWriter.java [hunk 11]
```diff
@@ -228,7 +169,7 @@ public void writeUTF8(byte[] string) {
   @Override
   public void writeUTF8(UTF8BytesString string) {
     writeStringHeader(string.encodedLength());
-    string.transferTo(buffer);
+    buffer.put(string.getUtf8Bytes());
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/msgpack/MsgPackWriter.java:172]: So what is the reason for this change? The resulting code seems to be the same to me, an `encodeIfNecessary()` followed by a `buffer.put(utf8Bytes)`.

Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/core/serialization/msgpack/MsgPackWriter.java:172]: Doh! Missed that.

File: dd-trace-core/src/main/java/datadog/trace/core/serialization/GrowableBuffer.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
+package datadog.trace.core.serialization;
+
+import java.nio.ByteBuffer;
+
+public final class GrowableBuffer implements StreamingBuffer {
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/serialization/GrowableBuffer.java:5]: Maybe add  the comment from this PR description here so it's clear when it should be used, and that it has exponential growth.

**Review by bantonsson — APPROVED** [review]

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java [hunk 13]
```diff
@@ -33,10 +29,14 @@
 
   private final WellKnownTags wellKnownTags;
   private final WritableFormatter writer;
+  private final Sink sink;
+  private final GrowableBuffer buffer;
 
   public SerializingMetricWriter(WellKnownTags wellKnownTags, Sink sink) {
     this.wellKnownTags = wellKnownTags;
-    this.writer = new MsgPackWriter(sink, ByteBuffer.allocate(1 << 20), EnumSet.of(SINGLE_MESSAGE));
+    this.buffer = new GrowableBuffer(512 << 10);
```
Comment (tylerbenson) [inline dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java:37]: In the doc for `GrowableBuffer` you say only use if bounded... what is it that makes metrics limited in size?

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy:203]: why no longer needed?

**Review by tylerbenson — APPROVED** [review]

File: dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy (hunk 6 above)
Comment (richardstartin) [reply dd-trace-core/src/test/groovy/datadog/trace/common/writer/DDAgentWriterCombinedTest.groovy:203]: because there is no header in the buffer any more, it's added when writing the bytes out to the network

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java (hunk 13 above)
Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/common/metrics/SerializingMetricWriter.java:37]: the metrics points are stored in a bounded LRU cache

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java (hunk 2 above)
Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/Payload.java:36]: this effectively _is_ a msgpack specific class and will be for the foreseeable future
